### PR TITLE
gh-117657: Add TSAN suppressions for the free-threaded build

### PR DESCRIPTION
**Pull request by mpage**
Add TSAN suppressions for existing races in the free-threaded build.

Additionally, skip a few tests that take a prohibitively long amount of time (> 1 hour) when TSAN is enabled and the GIL is disabled. I'm not sure why these are so much slower with TSAN (> 1 hour) vs without it (~1 minute). `perf` shows that all of the time is being spent in inclusively in GC in both configurations. When TSAN is enabled ~all of the exclusive cycles are in TSAN code (e.g `__tsan_read8`, `__tsan_read1`, etc.).

I cherry-picked this on top of the [nogil-integration branch](https://github.com/mpage/cpython/tree/nogil-integration) and verified that the free-threading TSAN tests pass locally and [in CI](https://github.com/mpage/cpython/actions/runs/8639220934/job/23685110519).

* Issue: gh-117657

### CODE REVIEW

File: Tools/tsan/supressions.txt
Comment (colesbury) [inline Tools/tsan/supressions.txt]: We should be able to remove `mi_heap_visit_pages` from here if we're splitting the free-threaded and default supressions.

File: Tools/tsan/suppressions_free_threading.txt
Comment (colesbury) [inline Tools/tsan/suppressions_free_threading.txt]: If we're categorizing them, the mimalloc suppressions are probably "free-threading specific".

**Review by colesbury — COMMENTED** [review]
Great!

Now that we have a list we can split up the work on fixing the data races.

**Comment by mpage** [comment]
> Now that we have a list we can split up the work on fixing the data races.

Yep! I'll create issues and link them to https://github.com/python/cpython/issues/117657.

**Review by corona10 — APPROVED** [review]
Well LGTM.

File: Lib/test/test_weakref.py
Comment (pitrou) [inline Lib/test/test_weakref.py]: Is it possible to just reduce the iteration count in this case instead? Ideally we shouldn't skip threaded stress tests on free-threading builds :-)

File: Tools/tsan/suppressions_free_threading.txt
Comment (pitrou) [reply Tools/tsan/suppressions_free_threading.txt]: Are these known by the mimalloc maintainers?

File: Tools/tsan/suppressions_free_threading.txt
Comment (pitrou) [inline Tools/tsan/suppressions_free_threading.txt:14]: Are there GH issues open for all these suppressions?

File: Lib/test/test_weakref.py
Comment (mpage) [reply Lib/test/test_weakref.py]: We have to reduce the iteration count by a couple orders of magnitude, but that should work.

File: Tools/tsan/suppressions_free_threading.txt
Comment (mpage) [reply Tools/tsan/suppressions_free_threading.txt:14]: Not yet, working on it :) I want to provide a bit more context (i.e. sample stacktraces + repro instructions) rather than just cutting issues without context.

File: Tools/tsan/suppressions_free_threading.txt
Comment (colesbury) [reply Tools/tsan/suppressions_free_threading.txt]: They are most likely issues in how we use mimalloc in the free-threading build rather than issues upstream.

File: Tools/tsan/suppressions_free_threading.txt
Comment (colesbury) [reply Tools/tsan/suppressions_free_threading.txt:14]: I'd like to start with a simple list in an omnibus issue like https://github.com/python/cpython/issues/117657 so that there's a single place that people can go look to track who is working on what races.

I expect many of the races to have simple fixes, but we can break off the more complex races into their own issues as needed.

The work of collecting stacktraces and repros is work that can be split up once this PR is landed and we have a way of coordinating the work.

Comment (mpage) [reply Tools/tsan/suppressions_free_threading.txt:14]: Sounds good. I threw together a quick script to massage the TSAN output from the tests into an [index](https://mpage.github.io/cpython_tsan_races/) that we can use to look for stacktraces and tests that should reproduce the races.

**Review by colesbury — APPROVED** [review]

**Pull request merged**
